Homework_1/report/main.tex

```diff
@@ -0,0 +1,141 @@
+\texttt{SimTridiag}, ki predstavlja simetrično tridiagonalno matriko, \texttt{Givens}, ki predstavlja matriko Givensovih rotacij in
+\texttt{ZgornjeTridiag}, ki predstavlja matriko z neničelnimi elementi na diagonali in dveh zgornjih diagonalah.
+
+Glavna naloga je tako implementacija funkcije \texttt{qr($T$)}, ki tridiagonalno matriko $T$ razcepi v produkt matrik $Q$ in $R$, kjer je $Q$ tipa \texttt{Givens} in $R$ tipa \texttt{ZgornjeTridiag} in
+implementacija funkcije \texttt{eigen($T$)}, ki izračuna lastne vrednosti in lastne vektorje simetrične tridiagonalne matrike $T$ z uporabo QR iteracije.
+
+\section{Implementacija}
+Podatkovni tip \texttt{Tridiag} sprejme tri sezname in sicer seznam spodnje diagonale, seznam glavne diagonale in seznam zgornje diagonale.
+Tale podatkovni tip smo že implementirali na vajah, zato je koda povzeta iz vaj. Podatkovni tip \texttt{SimTridiag} je podobno definiran, le da sprejme dva seznama, seznam glavne diagonale in seznam pod-diagonale, 
+saj sta zgornja in spodnja diagonala enaka. Podatkovni tip \texttt{ZgornjeTridiag} je podoben, le da sprejme seznam glavne diagonale in dva seznama zgornjih diagonal.
+
+Do elementov vseh treh matrik lahko dostopamo z indeksi, kot pri običajni matriki, kar smo storili z implementacijo funkcij \texttt{getindex()}. Elemente matrik lahko tudi spreminjamo s funkcijo \texttt{setindex()},
+vendar to lahko naredimo le za elemente, ki so v matriki definirani. V nasprotnem primeru, bo program vrnil napako.
+
+Pomemben podatkovni tip, ki si zasluži bolj podroben opis je \texttt{Givens}. Ta podatkovni sprejme seznam 4-dimenzionalnih vektorjev. Prvi dve komponenti sta kosinus in sinus kota rotacije, medtem ko sta tretja in četrta komponenta
+indeksa stolpcev/vrstic, ki jih rotiramo. Za boljšo predstavo si poglejmo matrično reprezentacijo Givensove rotacije
+\[
+G(c,s,i,j) =
+\begin{bmatrix}
+1      & \cdots & 0      & \cdots & 0      & \cdots & 0      \\
+\vdots & \ddots & \vdots &        & \vdots &        & \vdots \\
+0      & \cdots & c      & \cdots & -s     & \cdots & 0      \\
+\vdots &        & \vdots & \ddots & \vdots &        & \vdots \\
+0      & \cdots & s      & \cdots & c      & \cdots & 0      \\
+\vdots &        & \vdots &        & \vdots & \ddots & \vdots \\
+0      & \cdots & 0      & \cdots & 0      & \cdots & 1      
+\end{bmatrix}
+\]
+kjer se $c = \cos(\theta)$ in $s = \sin(\theta)$, kjer je $\theta$ kot rotacije, pojavita na presečišču $i$ in $j$-te vrstice in stolpca.
+
+Zadnji korak, ki ga moramo narediti preden se lotimo implementacije \texttt{qr($T$)}, je implementacija funkcije množenja med matrikami tipa \texttt{Givens} in \texttt{ZgornjeTridiag}. Ko bodisi z leve ali desne pomnožimo matriko tipa \texttt{ZgornjeTridiag}
+z matriko tipa \texttt{Givens}, ni nujno da je dobljena matrika tipa \texttt{ZgornjeTridiag}. Množenje teh dveh matrik smo zato implementirali tako, da smo matriko tipa \texttt{ZgornjeTridiag} najprej pretvorili v polno matriko ter
```

Comment:
Pri QR razcepu tridiagonalne matrike, bodo indeksi v matriki tipa `Givens` urejeni po vrsti in rezultat množenja bo ustrezal tipu `ZgornjeTridiag`.  Pretvorba v polno matriko nima smisla, glede na namen domače naloge. Kvečjemu, če uporabite razpršeno matriko. Bolje je pred množenjem preveriti, kakšni so indeksi v matriki tipa `Givens` in se nato odločiti, kakšen naj bo tip produkta. 

Še bolje je, da za tip `Givens` predpostavite, da so indeksi zaporedni in se izognete razvejišču v metodi množenja.